### PR TITLE
Provide a better error message for failed library loads on jdk17+

### PR DESCRIPTION
**Pull request by pshipton**
JVM_LoadLibrary() is used to load JNI libraries from jdk17.

Fixes https://github.com/eclipse-openj9/openj9/issues/17285

The library being loaded needs to exist to get the better message. With this change you get an error message to explain why the load failed.

`Exception in thread "main" java.lang.UnsatisfiedLinkError: Failed to load library ("/home/pete/openj9-openjdk-jdk17/build/linux-x86_64-server-release/images/jdk/lib/libabc.so") /home/pete/openj9-openjdk-jdk17/build/linux-x86_64-server-release/images/jdk/lib/libabc.so: file too short
`

`Exception in thread "main" java.lang.UnsatisfiedLinkError: Failed to load library ("/home/pete/openj9-openjdk-jdk17/build/linux-x86_64-server-release/images/jdk/lib/libabc.so") /home/pete/openj9-openjdk-jdk17/build/linux-x86_64-server-release/images/jdk/lib/libabc.so: invalid ELF header`

### CODE REVIEW

File: runtime/j9vm/jvm.c [hunk 1]
```diff
@@ -3968,12 +3969,15 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
 			UDATA handle = 0;
 			UDATA flags = J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags, J9_EXTENDED_RUNTIME_LAZY_SYMBOL_RESOLUTION) ? J9PORT_SLOPEN_LAZY : 0;
 			UDATA slOpenResult = j9sl_open_shared_library((char *)libName, &handle, flags);
-
 			Trc_SC_LoadLibrary_OpenShared(libName);
+			attemptedLoad = TRUE;
+
+#if JAVA_SPEC_VERSION < 17
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:3975]: I think this deserves a comment explaining why we don't want to try again for Java 17+.

File: runtime/j9vm/jvm.c [hunk 2]
```diff
@@ -3991,10 +3995,35 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
 		(*vm)->GetEnv(vm, (void **)&env, JNI_VERSION_1_2);
 		if (NULL != env) {
-			char errMsg[512];
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg = attemptedLoad ? j9error_last_error_message() : "";
+			if (0 != *portMsg) {
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4002]: Here and on line 4013, the constant should be `'\0'` (like line 4008).

File: runtime/j9vm/jvm.c [hunk 3]
```diff
@@ -3991,10 +3995,35 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
+						jio_snprintf(errMsg, bufSize, "Failed to load library %s", portMsg);
+					} else {
+						jio_snprintf(errMsg, bufSize, "Failed to load library \"%s\"", libName);
+					}
+				} else {
+					errMsg = errBuf;
+				}
+			}
 			throwNewUnsatisfiedLinkError(env, errMsg);
+			if (errMsg != errBuf) {
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4024]: Constant on left:
```c
			if (errBuf != errMsg) {
```

File: runtime/j9vm/jvm.c [hunk 4]
```diff
@@ -3991,10 +3995,35 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
+				j9mem_free_memory(errMsg);
+			}
 		}
 	}
 #endif /* JAVA_SPEC_VERSION >= 17 */
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4029]: The condition on line 3992 and here should be `JAVA_SPEC_VERSION >= 11`; see the function description.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
The description and commit message should be updated to reflect that this applies to jdk11+.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (pshipton) [reply runtime/j9vm/jvm.c:4029]: That's out of scope of my change. It was already JAVA_SPEC_VERSION >= 17 and 11 seems to work fine without it. Searching the OpenJDK 11 code, JVM_LoadLibrary is only used from ClassLoader.c, which isn't used with OpenJ9.

**Comment by pshipton** [comment]
> The description and commit message should be updated to reflect that this applies to jdk11+.

It only applied to jdk17+. jdk17 changed how JNI libraries are loaded, making this change necessary. jdk11 uses OpenJ9 to load libraries, which already provides the better error messages.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:4032]: Fair enough. Perhaps the function description could be improved to clarify that although the signature changed in jdk11, it is only used by jdk17+.

File: runtime/j9vm/jvm.c [hunk 5]
```diff
@@ -3968,12 +3969,18 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
 			UDATA handle = 0;
 			UDATA flags = J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags, J9_EXTENDED_RUNTIME_LAZY_SYMBOL_RESOLUTION) ? J9PORT_SLOPEN_LAZY : 0;
 			UDATA slOpenResult = j9sl_open_shared_library((char *)libName, &handle, flags);
-
 			Trc_SC_LoadLibrary_OpenShared(libName);
+			attemptedLoad = TRUE;
+
+/* jdk17+ calls JVM_LoadLibrary with decorated library names. If the following is done
+ * then it overwrites the real error with a failure to load "liblib<name>.so.so".
+ */
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:3977]: Actually, we don't expect this function to be used at all prior to jdk17; I'll plan to create a follow-up PR that simply asserts that and remove unnecessary code.

Comment (keithc-ca) [reply runtime/j9vm/jvm.c:3977]: We can use `defined(OPENJ9_BUILD)` to avoid impacting IBM Java 8 builds.

File: runtime/j9vm/jvm.c [hunk 6]
```diff
@@ -3991,10 +3998,35 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
 		if (NULL != env) {
-			char errMsg[512];
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg = attemptedLoad ? j9error_last_error_message() : "";
+			if ('\0' != *portMsg) {
+				/* The port library message includes the file name which failed to load. */
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4006]: Have you verified this on all platforms (the file name is part of the message)?

Comment (keithc-ca) [reply runtime/j9vm/jvm.c:4006]: The code in [`ClassLoader`](https://github.com/eclipse-openj9/openj9/blob/master/jcl/src/java.base/share/classes/java/lang/ClassLoader.java#L2009) creates a message which includes the file name as well as the error message which suggests not all platforms are has helpful as Linux.

`JVM_LoadSystemLibrary()` doesn't include the error message in the exception it throws (but I don't see any uses of that function (in OpenJ9 builds)).

Comment (pshipton) [reply runtime/j9vm/jvm.c]: AIX, Linux, Mac all include the name. z/OS does not, and I don't think Windows does either. I've pushed a change to include the name.

File: runtime/j9vm/jvm.c [hunk 7]
```diff
@@ -3991,10 +3998,25 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
 		(*vm)->GetEnv(vm, (void **)&env, JNI_VERSION_1_2);
 		if (NULL != env) {
-			char errMsg[512];
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg = attemptedLoad ? j9error_last_error_message() : "";
+			bufSize = jio_snprintf(errMsg, bufSize, "Failed to load library (\"%s\") %s", libName, portMsg);
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4005]: Perhaps a minor nit, but the resulting message will end with a space if `portMsg` is empty; maybe the we should keep the format alternatives?

Comment (keithc-ca) [reply runtime/j9vm/jvm.c:4005]: Perhaps like this?
```c
			const char *errFormat = ('\0' != *portMsg)
					? "Failed to load library (\"%s\") %s"
					: "Failed to load library (\"%s\")";
			bufSize = jio_snprintf(errMsg, bufSize, errFormat, libName, portMsg);
```

Comment (pshipton) [reply runtime/j9vm/jvm.c:4005]: It seems quite minor for something that would likely only occur in an OOM condition, it didn't seem worth the effort. But if you insist then I will fix it.

Comment (keithc-ca) [reply runtime/j9vm/jvm.c:4005]: Yes, perhaps unlikely to occur, but the delta is simple enough I think we should fix it.

Comment (pshipton) [reply runtime/j9vm/jvm.c]: Pushed

File: runtime/j9vm/jvm.c [hunk 8]
```diff
@@ -3991,10 +3998,26 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
 		(*vm)->GetEnv(vm, (void **)&env, JNI_VERSION_1_2);
 		if (NULL != env) {
-			char errMsg[512];
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg = attemptedLoad ? j9error_last_error_message() : "";
+			const char *space = ('\0' == *portMsg) ? "" : " ";
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4005]: Even better, I like it. Thanks.

**Comment by keithc-ca** [comment]
Jenkins compile win32 jdk8

**Comment by keithc-ca** [comment]
Jenkins compile alinux64 jdk17

**Comment by keithc-ca** [comment]
I'm not sure whether to expect the jdk8 build to fail because `attemptedLoad` is read only for jdk17+.

File: runtime/j9vm/jvm.c [hunk 9]
```diff
@@ -3991,10 +4002,30 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+#if defined(WIN32)
+			const char *portMsg = attemptedLoad ? j9error_last_error_message() : "";
+#else /* !defined(WIN32) */
+			const char *portMsg = j9error_last_error_message();
+#endif /* !defined(WIN32) */
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4012]: Please remove `!` from each comment.

File: runtime/j9vm/jvm.c [hunk 10]
```diff
@@ -3991,10 +4002,31 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg =
+#if defined(WIN32)
+			 !attemptedLoad ? "" :
+#else /* defined(WIN32) */
+			j9error_last_error_message();
+#endif /* defined(WIN32) */
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4013]: This is not syntactically valid for `WIN32`; (correcting indentation) we want:
```c
			const char *portMsg =
#if defined(WIN32)
				!attemptedLoad ? "" :
#endif /* defined(WIN32) */
				j9error_last_error_message();
```

File: runtime/j9vm/jvm.c [hunk 11]
```diff
@@ -3991,10 +4002,30 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
...
-			jio_snprintf(errMsg, sizeof(errMsg), "Failed to load library \"%s\"", libName);
-			errMsg[sizeof(errMsg) - 1] = '\0';
+			char errBuf[512];
+			char *errMsg = errBuf;
+			int bufSize = sizeof(errBuf);
+			const char *portMsg =
+#if defined(WIN32)
+			 !attemptedLoad ? "" :
+#endif /* defined(WIN32) */
+			j9error_last_error_message();
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c:4012]: Lines 4010 and 4012 should be indented one tab (and no spaces) more than line 4008.

**Comment by keithc-ca** [comment]
Jenkins compile win32 jdk8

**Comment by keithc-ca** [comment]
Jenkins compile amac jdk17

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**